### DISCUSSION TITLE
exploring-the-dotnet-8-preview-using-the-new-configuration-binder-source-generator/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#381] [Announcements]
# exploring-the-dotnet-8-preview-using-the-new-configuration-binder-source-generator/

In this post I introduce the new configuration binder source generator added in .NET 8 preview 3, explore how it works, and how to use it in your apps.

https://andrewlock.net/exploring-the-dotnet-8-preview-using-the-new-configuration-binder-source-generator/

**Comment by luizfbicalho** [comment]
I probably missed the point here, but is there a way to create the view model for the configuration based on the json, this way if I change the json config I get the new class changed automatically?

**Reply by shoaibshakeel381** [reply]
this source generator is only for getting config class/record properties and assigning their values by picking config values from dictionary. It won't create new properties in your code.

**Comment by shoaibshakeel381** [comment]
@andrewlock I tried to test method interception based on namespace and generic constraint criteria you mentioned in this post. But it does not seem to work for me. Consider following code:

```csharp
namespace OriginalCode
{
    using Extensions;

    internal sealed class Program
    {
        private static void Main()
        {
            var user = new User("London");
            user.PrintUserAddress();
        }
    }

    public record User(string City);
}

namespace Extensions
{
    public static class InnerUserExtensions
    {
        public static void PrintUserAddress(this global::OriginalCode.User user) => Console.WriteLine(user.City);
    }
}

public static class OuterUserExtensions
{
    public static void PrintUserAddress(this global::OriginalCode.User user) => Console.WriteLine("New York");
}
```

I have two extension methods on record type `User`. One is inside `Extensions` namespace and other is in global namespace. Note that in `OriginalCode` i've added a `using` statement for `Extensions` namespace. But I'm still seeing that extension method inside `InnerUserExtensions` class is executed.

If I remove  `using` statement for `Extensions` namespace, then extension method from `OuterUserExtensions` class will run, but that won't be that case when this binding source generator runs.

Edit Notes: simplified example

**Reply by shoaibshakeel381** [reply]
an update, seems like this trick only works if `using` statement for `Extensions` namespace is placed outside of `OriginalCode` namespace.

**Reply by khalidabuhakmeh** [reply]
I think it has more to do with the locality of the extension method. If an extension method is generated and added to your assembly, that is the first extension method found that matches the signature.

> When the compiler encounters a method invocation, it first looks for a match in the type's instance methods. If no match is found, it will search for any extension methods that are defined for the type, and bind to the first extension method that it finds.  [Microsoft Docs](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods#binding-extension-methods-at-compile-time)

I would think it scans the currently assembly first for an extension method, then external ones. Seeing that the original extension comes from the `Microsoft.Extensions.Configuration` package, it will always lose out to a similar extension method that originates from your current app.

I think the namespace is there to make the lives of developers easier, so they don't have to change anything.

**Reply by shoaibshakeel381** [reply]
Thnks for explanation, I do understand the resolution mechanism for which extension method will be picked. I was just wondering that since this trick will break, if I keep my `using` statements closer to the code, would it also break for `file-scoped using` statements? Also would that limit the sourcegenerator usefulness.

**Reply by andrewlock** [reply]
You know, it's interesting. It's not at all clear to me which extension is picked tbh.

I copied the generated code into the same project and played with the namespace, and looked to check which extension was chosen:

- with no namespace (default, global) - binds the custom extension
- with the `Microsoft.Extensions.DependencyInjection` namespace (or any other namespace in scope) - compiler error, ambiguous invocation
- with `global::Microsoft.Extensions.DependencyInjection` namespace - binds original extension (even if you use a `using`)

So in _general_ it doesn't seem like this holds:

> I would think it scans the currently assembly first for an extension method, then external ones. Seeing that the original extension comes from the Microsoft.Extensions.Configuration package, it will always lose out to a similar extension method that originates from your current app.

So, yeah, not sure exactly what's going on in your example! 😅

**Comment by VedhalamGanesh** [comment]
How's the mean standard deviation etc. calculated for above code? 
Kindly help.

**Comment by john--** [comment]
Do you have any examples of a .json file being loaded and enums being generated based off the config? Source Generators are so awkward to work with that it makes troubleshooting cumbersome.

**Reply by khalidabuhakmeh** [reply]
@john-- you should be able to debug source generators in most IDEs. I use JetBrains Rider. I wrote about it here. Hope that helps. https://blog.jetbrains.com/dotnet/2023/07/13/debug-source-generators-in-jetbrains-rider/

**Comment by jubba2001** [comment]
God bless you! I moved some .MapGet endpoints to a class library project (.NET 9). It worked fine until it was recompiled as Native-Aot. It was a day of agony before I reread your post and it gave me the right solution. Just need to add to class library project:

```
<PropertyGroup>
  <EnableRequestDelegateGenerator>true</EnableRequestDelegateGenerator>
  <EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator> 
  <InterceptorsNamespaces>$(InterceptorsNamespaces);Microsoft.AspNetCore.Http.Generated</InterceptorsNamespaces>
 
 <InterceptorsNamespaces>$(InterceptorsNamespaces);Microsoft.Extensions.Configuration.Binder.SourceGeneration</InterceptorsNamespaces>
</PropertyGroup>
```
